### ISSUE TITLE
Fix canonical/convenience split in NumericRecordDistribution

### ISSUE THREAD

**Issue by arob5** [TARPS-group/prob-pipe#175]
## Background

`NumericRecordDistribution` is intended to be the most general notion of a numeric random variable in ProbPipe: a distribution whose values have a `NumericRecord` structure (i.e., a pytree of `jax.Array` leaves, named via `RecordTemplate`). The single-leaf case (one named array — `Normal`, `Beta`, `MultivariateNormal`, etc.) is a trivial special case of this, and most of the class's logic should be reused unchanged across the trivial and structured cases.

This shape lines up cleanly with PR-C's "one random variable per `Distribution`" rule (#173, #174): the random variable is one entity, but its value can be a tree.

To make standard distributions usable in the typical way (`dist.dtype`, `dist.event_shape`, `dist.support`), the class also exposes singular convenience accessors alongside the general per-leaf ones (`dtypes`, `event_shapes`, `supports`). The intended pattern is:

| Concept | Canonical (per-leaf, tree-shaped) | Convenience (single-leaf shortcut) |
| --- | --- | --- |
| Structure | `record_template` | — |
| Pytree | `treedef` | — |
| Shapes | `event_shapes : dict[str, tuple]` | `event_shape : tuple[int, ...]` |
| Dtypes | `dtypes : dict[str, dtype]` | `dtype : dtype \| None` |
| Supports | `supports : dict[str, Constraint]` | `support : Constraint` |
| Total flat dim | `event_size : int` | — |

Convenience accessors should derive from canonical ones (raising or returning `None` when the tree has more than one leaf, matching the existing `RecordDistribution.shape` / `_single_field_name` precedent).

## The problem

The class advertises the multi-leaf interface but has several surfaces hardcoded for the single-leaf case, and subclasses override the convenience surface instead of (or in addition to) the canonical one — so the truth direction varies subclass-to-subclass and the two surfaces can drift.

Concretely:

### 1. `event_shape` is abstract; `event_shapes` is concrete on the parent

`event_shape` is the abstract method on `NumericRecordDistribution`, forcing every subclass to think of itself as single-leaf. `RecordDistribution.event_shapes` (the canonical per-leaf form) already exists and derives from `record_template`. Three subclasses (`TFPDistribution`, `BootstrapDistribution`, `FlattenedView`) each re-derive `event_shape` from a different source rather than letting it fall out of `record_template`.

### 2. `dtypes` default silently lies; `dtype` direction varies

The base `dtypes` returns `{name: _default_float_dtype() for name in tpl.fields}` — wrong for `Bernoulli`, `Poisson`, `Categorical`, etc. unless overridden. Compare to `supports`, which raises `NotImplementedError` and forces declaration.

The base `dtype` correctly derives from `dtypes` (unique value if all leaves agree, else `None`), but:
- `TFPDistribution` overrides both: `dtype` from the TFP dist, `dtypes` from `self.dtype` — flipping the canonical/convenience direction.
- `BootstrapDistribution` overrides only `dtype` and doesn't set a `record_template`, so `bd.dtype == float32` but `bd.dtypes == {}`. Strictly inconsistent.
- `FlattenedView` overrides neither; `dtype is None`, `dtypes == {}`.

### 3. `support` has the same shape problem as `dtype`, but no base convenience accessor

`NumericRecordDistribution.supports` is canonical (raises `NotImplementedError` if not overridden). There is no base-class `support` convenience — instead, `TFPDistribution.support` and `BootstrapDistribution.support` each define it ad-hoc.

### 4. `treedef` is a constant, not derived from `record_template`

```python
@property
def treedef(self) -> jax.tree_util.PyTreeDef:
    """Treedef for a single array (one-leaf pytree)."""
    return jax.tree.structure(None)
```

This is correct only when the sample is a raw array (the trivial single-leaf case). For a future multi-leaf subclass whose `_sample` returns a `NumericRecord`, the treedef must reflect that pytree shape. The relationship `treedef ↔ record_template` should be: `treedef` derives from `record_template` (leaf when single-leaf, NumericRecord skeleton otherwise).

### 5. `flat_event_shapes` is a vestigial `[self.event_shape]`

```python
@property
def flat_event_shapes(self) -> list[tuple[int, ...]]:
    """Single-leaf: just the one event_shape."""
    return [self.event_shape]
```

A list-of-one assumes single-leaf. The general definition is "list of leaf shapes in tree order," obtained by walking `event_shapes` / `record_template`.

### 6. The class docstring claims a `_sample` contract that isn't honored

The header says: "When `record_template` is set (named distribution), samples are wrapped as `Record`. Otherwise, raw arrays are returned for backward compatibility." But `TFPDistribution.record_template` is auto-set, and `TFPDistribution._sample` returns a raw array. The intended (and currently working) behavior is **Story A**: trivial single-leaf templates → `_sample` returns a raw array; multi-leaf templates → `_sample` returns a `NumericRecord`. The docstring should match.

## Proposed direction

Make every canonical/convenience split uniform:

1. **`record_template` is the single source of truth for structure.** Every `NumericRecordDistribution` subclass must set it. Helper for "single-field template from `name + event_shape`" (TFP already has this — promote to the base) covers the trivial case.

2. **Move `event_shape` from abstract to a concrete single-leaf convenience.** Define on the base as: return the unique shape iff there is exactly one leaf, else raise `TypeError` (matches `RecordDistribution.shape` / `_single_field_name` precedent). Subclasses stop overriding `event_shape`; they declare `record_template` and the convenience derives.

3. **`dtypes` becomes the canonical dtype surface.** Base `dtypes` raises `NotImplementedError` (matching `supports`); base `dtype` derives from `dtypes` (already correct). Subclasses override only `dtypes`. Drop the `dtype` overrides on `TFPDistribution` and `BootstrapDistribution`.

4. **Add a base-class `support` convenience accessor** derived from `supports` (raise on multi-leaf). Subclasses override only `supports`.

5. **Derive `treedef` from `record_template`** under Story A (single-leaf → leaf treedef; multi-leaf → `NumericRecord`-skeleton treedef). Functionally a no-op while only single-leaf concrete subclasses exist, but locks in the relationship.

6. **Replace `flat_event_shapes` with a tree-walk over `event_shapes`** (one entry per leaf, in tree order), or drop it if no callers need it.

7. **Tighten the class docstring** to describe the design intent: "Distribution over a `NumericRecord` (a pytree of `jax.Array` leaves). The trivial single-leaf case — a single named array — is the most common, and convenience accessors (`event_shape`, `dtype`, `support`) make it usable without thinking about tree structure."

8. **Apply the same canonical/convenience discipline to converter helpers** (`_default_support`, `_check_support_compatible`): under multi-leaf, support compatibility is a per-leaf check.

## Open question: should dtypes belong on `RecordTemplate`?

If `record_template` is going to be the single structural source of truth, an attractive option is to extend it to carry per-leaf dtypes (and possibly supports) as well. Then `dtypes` would derive from `record_template` automatically — no per-subclass override needed for the trivial case — and we'd have a single object that fully describes the random variable's structural type.

Arguments for:
- Single source of truth for structure *and* element type. Subclasses become very thin: declare a `record_template`, implement `_sample` / `_log_prob`, done.
- Symmetric with how `RecordTemplate.from_record` already inspects `.dtype` of leaves to classify them as numeric — the information is already being looked at, just not stored.
- Mirrors the structure–dtype pairing in numpy / xarray / pandas dtypes APIs.

Arguments against:
- `RecordTemplate` is currently a *structural* skeleton (field names + shapes + nesting). Adding dtypes blurs its purpose and may force unnecessary precision (e.g., what's the dtype of a `Categorical` whose support is integer but whose log-probs are float? what's the dtype of a `MixtureOfDiscrete`?).
- Templates are sometimes constructed from non-distribution sources (`from_record`, manual construction in tests) where the dtype isn't always meaningful.
- Distributions can change their reported dtype via JAX's x64 toggle; baking dtype into the template makes that interaction more delicate.

A middle path is to give `RecordTemplate` an optional `leaf_dtypes` field used when present (and falling back to subclass-defined `dtypes` otherwise), but that just defers the question.

Calling this out for discussion before deciding the direction in (3) above.

## Out of scope for this issue

- `_sample` return type ergonomics (Story A vs always-`NumericRecord`) — Story A matches current behavior and is the smaller change; revisit only if usability complaints accumulate.
- `__repr__.batch_shape` branch and `FlattenedView.batch_shape` `getattr` defensiveness — both become mechanically dead after PR-C.3 strips `batch_shape` from `Distribution[T]`.
- The unused `SupportsExpectation` import, the `Record as _Values` alias, and the `event_size` duplication with the parent — trivial cleanups, can land alongside (1) or independently.

## Files most affected

- `probpipe/core/_numeric_record_distribution.py` (the class)
- `probpipe/distributions/_tfp_base.py` (drop redundant `dtype` override; restructure `dtypes`)
- `probpipe/core/record.py` (only if dtypes move onto `RecordTemplate`)
- `probpipe/converters/_probpipe.py` (touchpoints for `_default_support` / `_check_support_compatible`)
